src/runtime/server/middleware/session/index.ts

```diff
@@ -81,11 +139,38 @@ const getSession = async (event: H3Event): Promise<null | Session> => {
     return null
   }
 
+  const runtimeConfig = useRuntimeConfig()
+  const sessionOptions = runtimeConfig.session.session
+
   // 3. Is the session not expired?
-  const sessionExpiryInSeconds = useRuntimeConfig().session.session.expiryInSeconds
+  const sessionExpiryInSeconds = sessionOptions.expiryInSeconds
   if (sessionExpiryInSeconds !== null) {
     const now = dayjs()
     if (now.diff(dayjs(session.createdAt), 'seconds') > sessionExpiryInSeconds) {
+      await deleteSession(event) // Cleanup old session data to avoid leaks
+      return null
+    }
+  }
+
+  // 4. Check for IP pinning logic
+  if (sessionOptions.ipPinning) {
+    const hashedIP = session.ip
+
+    // 4.1. (Should not happen) No IP address present in the session even though the flag is enabled
+    if (!hashedIP) {
+      await deleteSession(event) // Cleanup to avoid leaks (and properly recreate a session)
+      return null
+    }
+
+    // 4.2. Get request's IP
+    const requestIP = getRequestIpAddress(event)
+
+    // 4.3. Ensure pinning
+    const matches = await ipAddressesMatch(requestIP, hashedIP)
+    if (!matches) {
+      // 4.4. Report session-jacking attempt
+      // TODO: Report session-jacking attempt from requestIP
+      // NOTE: DO NOT DELETE SESSION HERE, this would mean we eliminate session-jacking, but users could delete others' sessions
       return null
     }
   }
```

Comment:
Nit: I would also move this into a separate `checkSessionIpAddress(event)` function which can throw an error. You can also refactor `checkSessionExpirationTime` to throw errors as well

```ts
try {
    if (sessionOptions.expiryInSeconds  !== null) {
        checkSessionExpirationTime(event);
    }
    if (sessionOptions.ipPinning) {
        checkSessionIpAddress(event);
    }
    return session
} catch {
    return null;
}